README.md
MongoDB Atlas Collection for Ansible
=========

[![ci-ansible-test](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/workflows/ansible-test/badge.svg)](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/actions?query=workflow%3Aansible-test)

This collection contains Ansible [modules](plugins/modules/) to provision and automate the cloud offering of MongoDB via API.

Required Ansible version: 2.9

Installation
------------

These modules are distributed as [collections](https://docs.ansible.com/ansible/latest/user_guide/collections_using.html).
To install them, run:

```
ansible-galaxy collection install t_systems_mms.mongodb_atlas
```

Alternatively put the collection into a `requirements.yml`-file:

```
---
collections:
- t_systems_mms.mongodb_atlas
```

Documentation
-------------

Our modules include documentation.

You can find the complete documentation for the modules in the [docs-folder](docs).

To display it on the command-line you can use the `ansible-doc` command.

License
-------

GPLv3

Author Information
------------------

* Martin Schurz


CHANGELOG.md
# Changelog

## [0.3.4](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/tree/0.3.4) (2023-06-15)

[Full Changelog](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/compare/0.3.3...0.3.4)

**Merged pull requests:**

- run CI-tests only once peer week [\#30](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/30) ([rndmh3ro](https://github.com/rndmh3ro))
- use reusable workflow [\#29](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/29) ([rndmh3ro](https://github.com/rndmh3ro))

## [0.3.3](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/tree/0.3.3) (2022-07-29)

[Full Changelog](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/compare/0.3.2...0.3.3)

**Fixed bugs:**

- API: handle 404 error [\#27](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/27) ([Raznak](https://github.com/Raznak))

**Merged pull requests:**

- update list of ansible and python versions to be tested [\#28](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/28) ([schurzi](https://github.com/schurzi))

## [0.3.2](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/tree/0.3.2) (2022-06-10)

[Full Changelog](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/compare/0.3.1...0.3.2)

**Implemented enhancements:**

- Allow new mongodb atlas versions [\#21](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/21) ([hack3d](https://github.com/hack3d))

**Merged pull requests:**

- Update actions/setup-python action to v4 [\#26](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/26) ([renovate[bot]](https://github.com/apps/renovate))
- Update github-actions-x/commit action to v2.9 [\#25](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/25) ([renovate[bot]](https://github.com/apps/renovate))
- Update charmixer/auto-changelog-action action to v1.4 [\#23](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/23) ([renovate[bot]](https://github.com/apps/renovate))
- Configure Renovate [\#20](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/20) ([renovate[bot]](https://github.com/apps/renovate))
- fix documentation workflow [\#19](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/19) ([schurzi](https://github.com/schurzi))
- add workflow job to autgenerate documentation [\#18](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/18) ([schurzi](https://github.com/schurzi))

## [0.3.1](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/tree/0.3.1) (2021-08-18)

[Full Changelog](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/compare/0.3.0...0.3.1)

**Implemented enhancements:**

- atlas\_user - set correct scope if empty parameter  [\#17](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/17) ([schurzi](https://github.com/schurzi))

## [0.3.0](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/tree/0.3.0) (2021-08-17)

[Full Changelog](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/compare/0.2.0...0.3.0)

**Implemented enhancements:**

- add scope to user, so permissions can be restricted to clusters [\#16](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/16) ([schurzi](https://github.com/schurzi))

**Merged pull requests:**

- Improve Release Action [\#15](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/15) ([schurzi](https://github.com/schurzi))
- use version for github action, short sha is no longer supported [\#14](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/14) ([schurzi](https://github.com/schurzi))
- fix galaxy-release action [\#13](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/13) ([rndmh3ro](https://github.com/rndmh3ro))
- update GitHub workflow [\#12](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/12) ([schurzi](https://github.com/schurzi))
- add doc\_fragment for global options [\#11](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/11) ([schurzi](https://github.com/schurzi))

## [0.2.0](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/tree/0.2.0) (2021-01-01)

[Full Changelog](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/compare/0.1.0...0.2.0)

**Implemented enhancements:**

- new module atlas\_ldap\_user for DB Users authenticated via LDAP [\#10](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/10) ([derekgiri](https://github.com/derekgiri))

**Closed issues:**

- Remove unwanted files from release-tarball  [\#5](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/issues/5)
- mixing of camel case and underscore variable names [\#2](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/issues/2)

**Merged pull requests:**

- Create runtime.yml [\#9](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/9) ([rndmh3ro](https://github.com/rndmh3ro))
- Create LICENSE [\#8](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/8) ([rndmh3ro](https://github.com/rndmh3ro))
- run tests on a schedule [\#7](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/7) ([rndmh3ro](https://github.com/rndmh3ro))
- add build\_ignore to filter unneeded files from release [\#6](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/pull/6) ([schurzi](https://github.com/schurzi))

## [0.1.0](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/tree/0.1.0) (2020-08-14)

[Full Changelog](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/compare/0.0.1...0.1.0)

**Implemented enhancements:**

- Use a loop to manage optional parameters [\#1](https://github.com/T-Systems-MMS/ansible-collection-mongodb-atlas/issues/1)

plugins/module_utils/atlas.py
# -*- coding: utf-8 -*-

from __future__ import absolute_import, division, print_function

__metaclass__ = type

import json
from collections import defaultdict

from ansible.module_utils.urls import fetch_url

try:
    from urllib import quote
except ImportError:
    # noinspection PyCompatibility, PyUnresolvedReferences
    from urllib.parse import (
        quote,
    )  # pylint: disable=locally-disabled, import-error, no-name-in-module


class AtlasAPIObject:
    module = None

    def __init__(
        self, module, object_name, groupId, path, data, data_is_array=False
    ):
        self.module = module
        self.path = path
        self.data = data
        self.groupId = groupId
        self.object_name = object_name
        self.data_is_array = data_is_array

        self.module.params["url_username"] = self.module.params["apiUsername"]
        self.module.params["url_password"] = self.module.params["apiPassword"]

    def call_url(self, path, data="", method="GET"):
        headers = {
            "Accept": "application/json",
            "Content-Type": "application/json",
        }

        if self.data_is_array and data != "":
            data = "[" + data + "]"

        url = (
            "https://cloud.mongodb.com/api/atlas/v1.0/groups/"
            + self.groupId
            + path
        )
        rsp, info = fetch_url(
            module=self.module,
            url=url,
            data=data,
            headers=headers,
            method=method,
        )

        content = ""
        error = ""
        if rsp and info["status"] != 204:
            content = json.loads(rsp.read())
        if info["status"] >= 400:
            try:
                content = json.loads(info["body"])
                error = content["reason"]
                if "detail" in content:
                    error += ". Detail: " + content["detail"]
            except ValueError:
                error = info["msg"]
        if info["status"] < 0:
            error = info["msg"]
        return {"code": info["status"], "data": content, "error": error}

    def exists(self):
        additional_path = ""
        if self.path == "/databaseUsers":
            additional_path = "/admin"
        ret = self.call_url(
            path=self.path
            + additional_path
            + "/"
            + quote(self.data[self.object_name], "")
        )
        if ret["code"] == 200:
            return True
        return False

    def create(self):
        ret = self.call_url(
            path=self.path,
            data=self.module.jsonify(self.data),
            method="POST",
        )
        return ret

    def delete(self):
        additional_path = ""
        if self.path == "/databaseUsers":
            additional_path = "/admin"
        ret = self.call_url(
            path=self.path
            + additional_path
            + "/"
            + quote(self.data[self.object_name], ""),
            method="DELETE",
        )
        return ret

    def modify(self):
        additional_path = ""
        if self.path == "/databaseUsers":
            additional_path = "/admin"
        ret = self.call_url(
            path=self.path
            + additional_path
            + "/"
            + quote(self.data[self.object_name], ""),
            data=self.module.jsonify(self.data),
            method="PATCH",
        )
        return ret

    def diff(self):
        additional_path = ""
        if self.path == "/databaseUsers":
            additional_path = "/admin"
        ret = self.call_url(
            path=self.path
            + additional_path
            + "/"
            + quote(self.data[self.object_name], ""),
            method="GET",
        )

        data_from_atlas = json.loads(self.module.jsonify(ret["data"]))
        data_from_task = json.loads(self.module.jsonify(self.data))

        diff = defaultdict(dict)
        for key, value in data_from_atlas.items():
            if key in data_from_task.keys() and value != data_from_task[key]:
                diff["before"][key] = "{val}".format(val=value)
                diff["after"][key] = "{val}".format(val=data_from_task[key])
        return diff

    def update(self, state):
        changed = False
        diff_result = {"before": "", "after": ""}
        if self.exists():
            diff_result.update({"before": "state: present\n"})
            if state == "absent":
                if self.module.check_mode:
                    diff_result.update({"after": "state: absent\n"})
                    self.module.exit_json(
                        changed=True,
                        object_name=self.data[self.object_name],
                        diff=diff_result,
                    )
                else:
                    try:
                        ret = self.delete()
                        if ret["code"] == 204 or ret["code"] == 202:
                            changed = True
                            diff_result.update({"after": "state: absent\n"})
                        else:
                            self.module.fail_json(
                                msg="bad return code while deleting: %d. Error message: %s"
                                % (ret["code"], ret["error"])
                            )
                    except Exception as e:
                        self.module.fail_json(
                            msg="exception when deleting: " + str(e)
                        )

            else:
                diff_result.update(self.diff())
                if self.module.check_mode:
                    if diff_result["after"] != "":
                        changed = True
                    self.module.exit_json(
                        changed=changed,
                        object_name=self.data[self.object_name],
                        data=self.data,
                        diff=diff_result,
                    )
                if diff_result["after"] != "":
                    if self.path == "/whitelist":
                        ret = self.create()
                    else:
                        ret = self.modify()
                    if ret["code"] == 200 or ret["code"] == 201:
                        changed = True
                    else:
                        self.module.fail_json(
                            msg="bad return code while modifying: %d. Error message: %s"
                            % (ret["code"], ret["error"])
                        )

        else:
            diff_result.update({"before": "state: absent\n"})
            if state == "present":
                if self.module.check_mode:
                    changed = True
                    diff_result.update({"after": "state: created\n"})
                else:
                    try:
                        ret = self.create()
                        if ret["code"] == 201:
                            changed = True
                            diff_result.update({"after": "state: created\n"})
                        else:
                            self.module.fail_json(
                                msg="bad return code while creating: %d. Error message: %s"
                                % (ret["code"], ret["error"])
                            )
                    except Exception as e:
                        self.module.fail_json(
                            msg="exception while creating: " + str(e)
                        )
        return changed, diff_result


plugins/modules/atlas_cluster.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
# Copyright (c) 2020 T-Systems MMS
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
#
# This module is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This software is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this software.  If not, see <http://www.gnu.org/licenses/>.

from __future__ import absolute_import, division, print_function

__metaclass__ = type

ANSIBLE_METADATA = {
    "metadata_version": "0.1",
    "status": ["preview"],
    "supported_by": "community",
}

DOCUMENTATION = """
---
module: atlas_cluster
short_description: Manage database clusters in Atlas
description:
  - The clusters module provides access to your cluster configurations.
  - The module lets you create, edit and delete clusters.
  - L(API Documentation,https://docs.atlas.mongodb.com/reference/api/clusters/)
author: "Martin Schurz (@schurzi)"
extends_documentation_fragment: t_systems_mms.mongodb_atlas.atlas_global_options
options:
  name:
    description:
      - Name of the cluster as it appears in Atlas. Once the cluster is created, its name cannot be changed.
    type: str
    required: True
  mongoDBMajorVersion:
    description:
      - Version of the cluster to deploy.
      - Atlas always deploys the cluster with the latest stable release of the specified version.
      - You can upgrade to a newer version of MongoDB when you modify a cluster.
    choices: [ "3.6", "4.0", "4.2", "4.4", "5.0", "6.0" ]
    type: str
  clusterType:
    description:
      - Type of the cluster that you want to create.
    choices: [ "REPLICASET", "SHARDED" ]
    default: "REPLICASET"
    type: str
  replicationFactor:
    description:
      - Number of replica set members. Each member keeps a copy of your databases, providing high availability and data redundancy.
    choices: [ 3, 5, 7 ]
    default: 3
    type: int
  autoScaling:
    description:
      - Configure your cluster to automatically scale its storage and cluster tier.
    suboptions:
      diskGBEnabled:
        type: bool
        description:
          - Specifies whether disk auto-scaling is enabled. The default is true.
    required: False
    type: dict
  providerSettings:
    description:
      - Configuration for the provisioned servers on which MongoDB runs.
      - The available options are specific to the cloud service provider.
    suboptions:
      providerName:
        required: True
        type: str
        description:
          - Cloud service provider on which the servers are provisioned.
      regionName:
        required: True
        type: str
        description:
          - Physical location of your MongoDB cluster.
      instanceSizeName:
        required: True
        type: str
        description:
          - Atlas provides different cluster tiers, each with a default storage capacity and RAM size.
          - The cluster you select is used for all the data-bearing servers in your cluster tier.
    required: True
    type: dict
  diskSizeGB:
    description:
      - Capacity, in gigabytes, of the host's root volume. Increase this number to add capacity,
        up to a maximum possible value of 4096 (i.e., 4 TB). This value must be a positive integer.
    type: int
  providerBackupEnabled:
    description:
      - Flag that indicates if the cluster uses Cloud Backups for backups.
    type: bool
  pitEnabled:
    description:
      - Flag that indicates the cluster uses continuous cloud backups.
    type: bool
"""

EXAMPLES = """
    - name: test cluster
      atlas_cluster:
        apiUsername: "API_user"
        apiPassword: "API_passwort_or_token"
        groupId: "GROUP_ID"
        name: "testcluster"
        mongoDBMajorVersion: "4.0"
        clusterType: "REPLICASET"
        providerSettings:
          providerName: "GCP"
          regionName: "EUROPE_WEST_3"
          instanceSizeName: "M10"
...
"""

from ansible.module_utils.basic import AnsibleModule
from ansible_collections.t_systems_mms.mongodb_atlas.plugins.module_utils.atlas import (
    AtlasAPIObject,
)


# ===========================================
# Module execution.
#
def main():
    # add our own arguments
    argument_spec = dict(
        state=dict(default="present", choices=["absent", "present"]),
        apiUsername=dict(required=True),
        apiPassword=dict(required=True, no_log=True),
        groupId=dict(required=True),
        name=dict(required=True),
        mongoDBMajorVersion=dict(
            choices=["3.6", "4.0", "4.2", "4.4", "5.0", "6.0"]
        ),
        clusterType=dict(
            default="REPLICASET", choices=["REPLICASET", "SHARDED"]
        ),
        replicationFactor=dict(default=3, type="int", choices=[3, 5, 7]),
        autoScaling=dict(
            type="dict",
            options=dict(
                diskGBEnabled=dict(type="bool"),
            ),
        ),
        providerSettings=dict(
            type="dict",
            required=True,
            options=dict(
                providerName=dict(required=True),
                regionName=dict(required=True),
                instanceSizeName=dict(required=True),
            ),
        ),
        diskSizeGB=dict(type="int"),
        providerBackupEnabled=dict(type="bool"),
        pitEnabled=dict(type="bool"),
    )

    # Define the main module
    module = AnsibleModule(
        argument_spec=argument_spec, supports_check_mode=True
    )

    data = {
        "name": module.params["name"],
        "clusterType": module.params["clusterType"],
        "replicationFactor": module.params["replicationFactor"],
        "providerSettings": module.params["providerSettings"],
    }

    # handle optional options
    optional_vars = [
        "mongoDBMajorVersion",
        "autoScaling",
        "diskSizeGB",
        "providerBackupEnabled",
        "pitEnabled",
    ]

    for var in optional_vars:
        if var in module.params:
            data.update({var: module.params[var]})

    try:
        atlas = AtlasAPIObject(
            module=module,
            path="/clusters",
            object_name="name",
            groupId=module.params["groupId"],
            data=data,
        )
    except Exception as e:
        module.fail_json(
            msg="unable to connect to Atlas API. Exception message: %s" % e
        )

    changed, diff = atlas.update(module.params["state"])
    module.exit_json(
        changed=changed,
        data=atlas.data,
        diff=diff,
    )


# import module snippets
if __name__ == "__main__":
    main()


plugins/modules/atlas_ldap_user.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
# Copyright (c) 2020 T-Systems MMS
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
#
# This module is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This software is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this software.  If not, see <http://www.gnu.org/licenses/>.

from __future__ import absolute_import, division, print_function

__metaclass__ = type

ANSIBLE_METADATA = {
    "metadata_version": "0.1",
    "status": ["preview"],
    "supported_by": "community",
}

DOCUMENTATION = """
---
module: atlas_ldap_user
short_description: Manage database users in Atlas
description:
  - The atlas_ldap_user module lets you create LDAP groups on the admin database by mapping LDAP groups to MongoDB roles on your Atlas databases.
  - Each user or group has a set of roles that provide access to the project's databases.
  - L(API Documentation,https://docs.atlas.mongodb.com/security-ldaps/)
author: "Martin Schurz (@schurzi) / Derek Giri"
extends_documentation_fragment: t_systems_mms.mongodb_atlas.atlas_global_options
options:
  databaseName:
    description:
      - Database against which Atlas authenticates the user.
    choices: ["admin", "$external"]
    default: "admin"
    type: str
  ldapAuthType:
    description:
      - Type of LDAP authorization for the user i.e. USER or GROUP
    choices: ["GROUP", "USER"]
    default: "GROUP"
    type: str
  username:
    description:
      - Username for authenticating to MongoDB.
    required: true
    type: str
  roles:
    description:
      - Array of this user's roles and the databases / collections on which the roles apply.
      - A role must include folliwing elements
    suboptions:
      databaseName:
        required: true
        type: str
        description:
          - Database on which the user has the specified role.
          - A role on the admin database can include privileges that apply to the other databases.
      roleName:
        required: true
        type: str
        description:
          - Name of the role. This value can either be a built-in role or a custom role.
    required: true
    type: list
    elements: dict

"""
EXAMPLES = """
    - name: LDAP Group or Username
      atlas_ldap_user:
        apiUsername: "API_user"
        apiPassword: "API_passwort_or_token"
        atlas_ldap_user: "USER DN or GROUP DN"
        groupId: "GROUP_ID"
        databaseName: "admin"
        username: my_app_user
        roles:
          - databaseName: private_info
            roleName: read
          - databaseName: public_info
            roleName: readWrite
"""

from ansible.module_utils.basic import AnsibleModule
from ansible_collections.t_systems_mms.mongodb_atlas.plugins.module_utils.atlas import (
    AtlasAPIObject,
)


# ===========================================
# Module execution.
#
def main():
    # add our own arguments
    argument_spec = dict(
        state=dict(default="present", choices=["absent", "present"]),
        apiUsername=dict(required=True),
        apiPassword=dict(required=True, no_log=True),
        groupId=dict(required=True),
        ldapAuthType=dict(default="GROUP", choices=["GROUP", "USER"]),
        databaseName=dict(default="admin", choices=["admin", "$external"]),
        username=dict(required=True),
        roles=dict(
            required=True,
            type="list",
            elements="dict",
            options=dict(
                databaseName=dict(required=True),
                roleName=dict(required=True),
            ),
        ),
    )

    # Define the main module
    module = AnsibleModule(
        argument_spec=argument_spec, supports_check_mode=True
    )

    data = {
        "databaseName": module.params["databaseName"],
        "ldapAuthType": module.params["ldapAuthType"],
        "username": module.params["username"],
        "roles": module.params["roles"],
    }

    try:
        atlas = AtlasAPIObject(
            module=module,
            path="/databaseUsers",
            object_name="username",
            groupId=module.params["groupId"],
            data=data,
        )
    except Exception as e:
        module.fail_json(
            msg="unable to connect to Atlas API. Exception message: %s" % e
        )

    changed, diff = atlas.update(module.params["state"])
    module.exit_json(
        changed=changed,
        data=atlas.data,
        diff=diff,
    )


# import module snippets
if __name__ == "__main__":
    main()


plugins/modules/atlas_whitelist.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
# Copyright (c) 2020 T-Systems MMS
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
#
# This module is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This software is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this software.  If not, see <http://www.gnu.org/licenses/>.

from __future__ import absolute_import, division, print_function

__metaclass__ = type

ANSIBLE_METADATA = {
    "metadata_version": "0.1",
    "status": ["preview"],
    "supported_by": "community",
}

DOCUMENTATION = """
---
module: atlas_whitelist
short_description: Manage IP whitelists in Atlas
description:
  - The whitelist module manages a Atlas project's IP whitelist.
  - L(API Documentation,https://docs.atlas.mongodb.com/reference/api/whitelist/)
author: "Martin Schurz (@schurzi)"
extends_documentation_fragment: t_systems_mms.mongodb_atlas.atlas_global_options
options:
  cidrBlock:
    description:
      - Whitelist entry in Classless Inter-Domain Routing (CIDR) notation.
    type: str
    required: True
  comment:
    description:
      - Optional Comment associated with the whitelist entry.
    type: str
    default: "created by Ansible"
"""

EXAMPLES = """
    - name: test whitelist
      atlas_whitelist:
        apiUsername: "API_user"
        apiPassword: "API_passwort_or_token"
        groupId: "GROUP_ID"
        cidrBlock: "192.168.0.0/24"
        comment: "test"
"""

from ansible.module_utils.basic import AnsibleModule
from ansible_collections.t_systems_mms.mongodb_atlas.plugins.module_utils.atlas import (
    AtlasAPIObject,
)


# ===========================================
# Module execution.
#
def main():
    # add our own arguments
    argument_spec = dict(
        state=dict(default="present", choices=["absent", "present"]),
        apiUsername=dict(required=True),
        apiPassword=dict(required=True, no_log=True),
        groupId=dict(required=True),
        cidrBlock=dict(required=True),
        comment=dict(default="created by Ansible"),
    )

    # Define the main module
    module = AnsibleModule(
        argument_spec=argument_spec, supports_check_mode=True
    )

    data = {
        "cidrBlock": module.params["cidrBlock"],
        "comment": module.params["comment"],
    }

    try:
        atlas = AtlasAPIObject(
            module=module,
            path="/whitelist",
            object_name="cidrBlock",
            groupId=module.params["groupId"],
            data=data,
            data_is_array=True,
        )
    except Exception as e:
        module.fail_json(
            msg="unable to connect to Atlas API. Exception message: %s" % e
        )

    changed, diff = atlas.update(module.params["state"])
    module.exit_json(
        changed=changed,
        data=atlas.data,
        diff=diff,
    )


# import module snippets
if __name__ == "__main__":
    main()


plugins/modules/atlas_user.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
# Copyright (c) 2020 T-Systems MMS
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
#
# This module is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This software is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this software.  If not, see <http://www.gnu.org/licenses/>.

from __future__ import absolute_import, division, print_function

__metaclass__ = type

ANSIBLE_METADATA = {
    "metadata_version": "0.1",
    "status": ["preview"],
    "supported_by": "community",
}

DOCUMENTATION = """
---
module: atlas_user
short_description: Manage database users in Atlas
description:
  - The atlas_users module lets you create, modify and delete the database users in your cluster.
  - Each user has a set of roles that provide access to the project's databases.
  - A user's roles apply to all the clusters in the project
  - if two clusters have a products database and a user has a role granting read access on the products database,
  - the user has that access on both clusters.
  - L(API Documentation,https://docs.atlas.mongodb.com/reference/api/database-users/)
author: "Martin Schurz (@schurzi)"
extends_documentation_fragment: t_systems_mms.mongodb_atlas.atlas_global_options
options:
  databaseName:
    description:
      - Database against which Atlas authenticates the user.
    choices: ["admin", "$external"]
    default: "admin"
    type: str
  username:
    description:
      - Username for authenticating to MongoDB.
    required: true
    type: str
  password:
    description:
      - User's password.
    required: true
    type: str
  roles:
    description:
      - Array of this user's roles and the databases / collections on which the roles apply.
      - A role must include following elements
    suboptions:
      databaseName:
        required: true
        type: str
        description:
          - Database on which the user has the specified role.
          - A role on the admin database can include privileges that apply to the other databases.
      roleName:
        required: true
        type: str
        description:
          - Name of the role. This value can either be a built-in role or a custom role.
    required: true
    type: list
    elements: dict
  scopes:
    description:
      - List of clusters and Atlas Data Lakes that this user can access.
      - Atlas grants database users access to all resources by default.
    suboptions:
      name:
        required: true
        type: str
        description:
          - Name of the cluster or Atlas Data Lake that the database user can access.
      type:
        type: str
        choices: ["CLUSTER", "DATA_LAKE"]
        default: "CLUSTER"
        description:
          - Type of resource that the database user can access.
    required: false
    default: []
    type: list
    elements: dict
"""

EXAMPLES = """
    - name: test user
      atlas_user:
        apiUsername: "API_user"
        apiPassword: "API_passwort_or_token"
        groupId: "GROUP_ID"
        username: my_app_user
        password: SuperSecret!
        roles:
          - databaseName: private_info
            roleName: read
          - databaseName: public_info
            roleName: readWrite
"""

from ansible.module_utils.basic import AnsibleModule
from ansible_collections.t_systems_mms.mongodb_atlas.plugins.module_utils.atlas import (
    AtlasAPIObject,
)


# ===========================================
# Module execution.
#
def main():
    # add our own arguments
    argument_spec = dict(
        state=dict(default="present", choices=["absent", "present"]),
        apiUsername=dict(required=True),
        apiPassword=dict(required=True, no_log=True),
        groupId=dict(required=True),
        databaseName=dict(default="admin", choices=["admin", "$external"]),
        username=dict(required=True),
        password=dict(required=True, no_log=True),
        roles=dict(
            required=True,
            type="list",
            elements="dict",
            options=dict(
                databaseName=dict(required=True),
                roleName=dict(required=True),
            ),
        ),
        scopes=dict(
            required=False,
            type="list",
            elements="dict",
            options=dict(
                name=dict(required=True),
                type=dict(default="CLUSTER", choices=["CLUSTER", "DATA_LAKE"]),
            ),
            default=[],
        ),
    )

    # Define the main module
    module = AnsibleModule(
        argument_spec=argument_spec, supports_check_mode=True
    )

    data = {
        "databaseName": module.params["databaseName"],
        "username": module.params["username"],
        "password": module.params["password"],
        "roles": module.params["roles"],
        "scopes": module.params["scopes"],
    }

    try:
        atlas = AtlasAPIObject(
            module=module,
            path="/databaseUsers",
            object_name="username",
            groupId=module.params["groupId"],
            data=data,
        )
    except Exception as e:
        module.fail_json(
            msg="unable to connect to Atlas API. Exception message: %s" % e
        )

    changed, diff = atlas.update(module.params["state"])
    module.exit_json(
        changed=changed,
        data=atlas.data,
        diff=diff,
    )


# import module snippets
if __name__ == "__main__":
    main()
